File: scripts/hard_priced_interval.py
"""Exact offline priced interval reference with hard GET and unit caps.

This reference retains the best nonnegative integer modeled mass for each
exact (GET, unit, open/closed) state. Mandatory units cannot be skipped.
The implementation is O(sites × max_gets × max_units) in time and stores
two trace bitmaps per site; a low-latency Rust refinement is separate work.
The weights are caller-supplied predictions, never ground truth here.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping, Sequence

import numpy as np


@dataclass(frozen=True)
class HardPricedCover:
    intervals: tuple[tuple[int, int], ...]
    mass: int
    units: int
    gets: int
    objective: int


def hard_priced_cover(
    weight_by_unit: Mapping[int, int], mandatory_units: Sequence[int], *,
    page_count: int, max_gets: int, max_units: int,
    unit_price: int, get_price: int,
) -> HardPricedCover:
    """Maximize modeled mass minus prices among physically feasible plans."""
    mandatory = tuple(mandatory_units)
    if (type(page_count) is not int or page_count <= 0
            or type(max_gets) is not int or max_gets <= 0
            or type(max_units) is not int or max_units <= 0
            or type(unit_price) is not int or unit_price < 0
            or type(get_price) is not int or get_price < 0
            or len(set(mandatory)) != len(mandatory)
            or any(type(unit) is not int or not 0 <= unit < page_count
                   for unit in mandatory)
            or any(type(unit) is not int or not 0 <= unit < page_count
                   or type(weight) is not int or weight <= 0
                   for unit, weight in weight_by_unit.items())
            or sum(weight_by_unit.values()) >= 2**30):
        raise ValueError("hard priced interval geometry differs")
    sites = sorted(set(mandatory) | set(weight_by_unit))
    if not sites:
        return HardPricedCover((), 0, 0, 0, 0)
    gets_cap = min(max_gets, len(sites))
    shape = (gets_cap + 1, max_units + 1)
    negative = -(2**40)
    closed = np.full(shape, negative, dtype=np.int64)
    opened = np.full(shape, negative, dtype=np.int64)
    closed[0, 0] = 0
    history: list[tuple[int, int, np.ndarray, np.ndarray, bool]] = []
    required = set(mandatory)
    previous: int | None = None
    for site in sites:
        weight = weight_by_unit.get(site, 0)
        continuation = 1 if previous is None else site - previous
        closed_from_open = opened > closed
        base = np.maximum(closed, opened)
        next_closed = (np.full(shape, negative, dtype=np.int64)
                       if site in required else base)
        next_open = np.full(shape, negative, dtype=np.int64)
        continued = np.zeros(shape, dtype=bool)
        next_open[1:, 1:] = base[:-1, :-1] + weight
        if continuation <= max_units:
            candidate = opened[:, :max_units + 1 - continuation] + weight
            section = next_open[:, continuation:]
            continued[:, continuation:] = candidate > section
            np.maximum(section, candidate, out=section)
        history.append((site, continuation, closed_from_open,
                        continued, site in required))
        closed, opened = next_closed, next_open
        previous = site

    best: tuple[int, int, int, int] | None = None
    best_open = False
    for gets, units in np.argwhere(np.maximum(closed, opened) >= 0):
        gets, units = int(gets), int(units)
        is_open = bool(opened[gets, units] > closed[gets, units])
        mass = int(max(closed[gets, units], opened[gets, units]))
        key = (mass - unit_price * units - get_price * gets,
               -units, -gets, mass)
        if best is None or key > best:
            best, best_open = key, is_open
    if best is None:
        raise ValueError("mandatory cover infeasible within hard caps")

    objective, neg_units, neg_gets, expected_mass = best
    cursor_units, cursor_gets = -neg_units, -neg_gets
    intervals: list[tuple[int, int]] = []
    pending_end: int | None = None
    for site, continuation, closed_from_open, continued, required_site in reversed(history):
        if not best_open:
            if required_site:
                raise AssertionError("hard priced mandatory site was skipped")
            best_open = bool(closed_from_open[cursor_gets, cursor_units])
            continue
        if pending_end is None:
            pending_end = site
        if continued[cursor_gets, cursor_units]:
            cursor_units -= continuation
        else:
            intervals.append((site, pending_end))
            pending_end = None
            cursor_gets -= 1
            cursor_units -= 1
            best_open = bool(closed_from_open[cursor_gets, cursor_units])
    if pending_end is not None or cursor_units or cursor_gets or best_open:
        raise AssertionError("hard priced interval witness reconstruction differs")
    intervals.reverse()
    witness = tuple(intervals)
    gets = len(witness)
    units = sum(end - start + 1 for start, end in witness)
    mass = sum(weight for site, weight in weight_by_unit.items()
               if any(start <= site <= end for start, end in witness))
    if (units > max_units or gets > max_gets
            or any(not any(start <= site <= end for start, end in witness)
                   for site in mandatory)
            or any(left[1] >= right[0]
                   for left, right in zip(witness, witness[1:]))
            or mass != expected_mass
            or objective != mass - unit_price * units - get_price * gets):
        raise AssertionError("hard priced interval witness accounting differs")
    return HardPricedCover(witness, mass, units, gets, objective)

File: crates/borsuk/src/hard_priced_interval.rs
//! Exact priced interval cover under simultaneous hard GET and unit caps.
//!
//! The two-dimensional fallback is used only when cheaper admitted optima
//! cannot certify the two-cap optimum. Its weights are caller predictions.

use std::collections::{BTreeMap, BTreeSet};

use crate::unconstrained_priced_interval::{CoverError, UnconstrainedCover};

struct Trace {
    site: usize,
    gap: usize,
    required: bool,
    closed_from_open: Vec<bool>,
    continued: Vec<bool>,
}

fn index(gets: usize, units: usize, stride: usize) -> usize {
    gets * stride + units
}

/// Maximize modeled mass minus fixed unit and GET prices under both caps.
///
/// Runtime is `O(sites * get_cap * unit_cap)` and the retained trace has
/// `2 * sites * (get_cap + 1) * (unit_cap + 1)` Boolean entries.
pub fn hard_priced_cover(
    weights: &BTreeMap<usize, i64>,
    mandatory: &[usize],
    page_count: usize,
    get_cap: usize,
    unit_cap: usize,
    unit_price: i64,
    get_price: i64,
    trace_budget_bytes: usize,
) -> Result<UnconstrainedCover, CoverError> {
    if page_count == 0
        || page_count > i64::MAX as usize
        || get_cap == 0
        || unit_cap == 0
        || get_cap > page_count
        || unit_cap > page_count
        || trace_budget_bytes == 0
        || unit_price < 0
        || get_price < 0
        || unit_price
            .checked_add(get_price)
            .and_then(|sum| sum.checked_mul(page_count as i64))
            .is_none_or(|charge| charge > i64::MAX - (1 << 30))
    {
        return Err(CoverError::InvalidGeometry);
    }
    let required = mandatory.iter().copied().collect::<BTreeSet<_>>();
    if required.len() != mandatory.len()
        || required.iter().any(|site| *site >= page_count)
        || weights
            .iter()
            .any(|(&site, &weight)| site >= page_count || weight <= 0)
    {
        return Err(CoverError::InvalidGeometry);
    }
    let mut total_mass = 0_i64;
    for &weight in weights.values() {
        total_mass = total_mass
            .checked_add(weight)
            .ok_or(CoverError::ArithmeticOverflow)?;
    }
    if total_mass >= 1 << 30 {
        return Err(CoverError::InvalidGeometry);
    }
    let sites = required
        .iter()
        .copied()
        .chain(weights.keys().copied())
        .collect::<BTreeSet<_>>();
    if sites.is_empty() {
        return Ok(UnconstrainedCover {
            intervals: Vec::new(),
            mass: 0,
            units: 0,
            gets: 0,
            objective: 0,
        });
    }
    let stride = unit_cap
        .checked_add(1)
        .ok_or(CoverError::ArithmeticOverflow)?;
    let slots = get_cap
        .checked_add(1)
        .and_then(|value| value.checked_mul(stride))
        .ok_or(CoverError::ArithmeticOverflow)?;
    let trace_bytes = sites
        .len()
        .checked_mul(slots)
        .and_then(|value| value.checked_mul(2))
        .ok_or(CoverError::ArithmeticOverflow)?;
    if trace_bytes > trace_budget_bytes {
        return Err(CoverError::TraceBudgetExceeded);
    }
    // All reachable modeled masses are nonnegative, so -1 is unreachable.
    let mut closed = vec![-1_i64; slots];
    let mut opened = vec![-1_i64; slots];
    closed[0] = 0;
    let mut history = Vec::with_capacity(sites.len());
    let mut previous = None;
    for site in sites {
        let weight = *weights.get(&site).unwrap_or(&0);
        let gap = previous.map_or(1, |old| site - old);
        let required_here = required.contains(&site);
        let mut next_closed = vec![-1_i64; slots];
        let mut next_open = vec![-1_i64; slots];
        let mut closed_from_open = vec![false; slots];
        let mut continued = vec![false; slots];
        for gets in 0..=get_cap {
            for units in 0..=unit_cap {
                let slot = index(gets, units, stride);
                let from_open = opened[slot] > closed[slot];
                closed_from_open[slot] = from_open;
                let base = closed[slot].max(opened[slot]);
                if !required_here {
                    next_closed[slot] = base;
                }
                if base >= 0 && gets < get_cap && units < unit_cap {
                    let target = index(gets + 1, units + 1, stride);
                    let candidate = base + weight;
                    if candidate > next_open[target] {
                        next_open[target] = candidate;
                        continued[target] = false;
                    }
                }
            }
        }
        // Match the reference's strict tie rule: all new intervals are
        // installed before continuations, so an equal-mass continuation
        // preserves the new-interval trace.
        if gap <= unit_cap {
            for gets in 0..=get_cap {
                for units in 0..=unit_cap - gap {
                    let slot = index(gets, units, stride);
                    if opened[slot] < 0 {
                        continue;
                    }
                    let target = index(gets, units + gap, stride);
                    let candidate = opened[slot] + weight;
                    if candidate > next_open[target] {
                        next_open[target] = candidate;
                        continued[target] = true;
                    }
                }
            }
        }
        history.push(Trace {
            site,
            gap,
            required: required_here,
            closed_from_open,
            continued,
        });
        closed = next_closed;
        opened = next_open;
        previous = Some(site);
    }
    let mut best: Option<(i64, i64, i64, i64)> = None;
    let mut best_gets = 0;
    let mut best_units = 0;
    let mut best_open = false;
    for gets in 0..=get_cap {
        for units in 0..=unit_cap {
            let slot = index(gets, units, stride);
            let mass = closed[slot].max(opened[slot]);
            if mass < 0 {
                continue;
            }
            let key = (
                mass - unit_price * units as i64 - get_price * gets as i64,
                -(units as i64),
                -(gets as i64),
                mass,
            );
            if best.is_none_or(|prior| key > prior) {
                best = Some(key);
                best_gets = gets;
                best_units = units;
                best_open = opened[slot] > closed[slot];
            }
        }
    }
    let winner = best.ok_or(CoverError::NoFeasibleCover)?;
    let mut gets = best_gets;
    let mut units = best_units;
    let mut pending_end = None;
    let mut intervals = Vec::new();
    for trace in history.into_iter().rev() {
        let slot = index(gets, units, stride);
        if !best_open {
            if trace.required {
                return Err(CoverError::InvalidWitness);
            }
            best_open = trace.closed_from_open[slot];
        } else {
            if pending_end.is_none() {
                pending_end = Some(trace.site);
            }
            if trace.continued[slot] {
                units = units
                    .checked_sub(trace.gap)
                    .ok_or(CoverError::InvalidWitness)?;
            } else {
                intervals.push((
                    trace.site,
                    pending_end.take().ok_or(CoverError::InvalidWitness)?,
                ));
                gets = gets.checked_sub(1).ok_or(CoverError::InvalidWitness)?;
                units = units.checked_sub(1).ok_or(CoverError::InvalidWitness)?;
                best_open = trace.closed_from_open[index(gets, units, stride)];
            }
        }
    }
    if gets != 0 || units != 0 || best_open || pending_end.is_some() {
        return Err(CoverError::InvalidWitness);
    }
    intervals.reverse();
    let actual_units = intervals.iter().try_fold(0_usize, |sum, &(start, end)| {
        sum.checked_add(end - start + 1)
            .ok_or(CoverError::ArithmeticOverflow)
    })?;
    let actual_gets = intervals.len();
    let mut mass = 0_i64;
    let mut interval = 0_usize;
    for (&site, &weight) in weights {
        while interval < intervals.len() && intervals[interval].1 < site {
            interval += 1;
        }
        if interval < intervals.len() && intervals[interval].0 <= site {
            mass += weight;
        }
    }
    let objective = mass - unit_price * actual_units as i64 - get_price * actual_gets as i64;
    if (
        objective,
        -(actual_units as i64),
        -(actual_gets as i64),
        mass,
    ) != winner
        || actual_units > unit_cap
        || actual_gets > get_cap
        || required
            .iter()
            .any(|site| !intervals.iter().any(|(a, b)| a <= site && site <= b))
    {
        return Err(CoverError::InvalidWitness);
    }
    Ok(UnconstrainedCover {
        intervals,
        mass,
        units: actual_units,
        gets: actual_gets,
        objective,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hard_caps_force_bridge_of_mandatory_sites() {
        let weights = BTreeMap::from([(0, 10), (2, 8)]);
        let cover = hard_priced_cover(&weights, &[0, 2], 3, 1, 3, 1, 3, 1024).unwrap();
        assert_eq!(cover.intervals, vec![(0, 2)]);
        assert_eq!((cover.mass, cover.units, cover.gets), (18, 3, 1));
    }

    #[test]
    fn infeasible_hard_caps_are_rejected() {
        let cover = hard_priced_cover(&BTreeMap::new(), &[0, 2], 3, 1, 2, 1, 3, 1024);
        assert_eq!(cover, Err(CoverError::NoFeasibleCover));
    }
}
